### ISSUE TITLE
[data grid] Event Propagation and Portal Challenges with DataGrid's GridActionsCellItem

### ISSUE THREAD

**Issue by irontitan76** [mui/mui-x#11579]
### Steps to reproduce

[Link to live example](https://codesandbox.io/p/devbox/rcdwnv?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqzq7roc00063b6hqu5oq7oy%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqzq7roc00023b6hpgno3t5g%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqzq7roc00033b6h868p1gcm%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqzq7roc00053b6h4c20jjsy%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqzq7roc00023b6hpgno3t5g%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqzq7rob00013b6hrt7gaex5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpublic%252Findex.html%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clqzq7roc00023b6hpgno3t5g%2522%252C%2522activeTabId%2522%253A%2522clqzq7rob00013b6hrt7gaex5%2522%257D%252C%2522clqzq7roc00053b6h4c20jjsy%2522%253A%257B%2522id%2522%253A%2522clqzq7roc00053b6h4c20jjsy%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clqzqa64300933b6hjavljurv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqzqa64300933b6hjavljurv%2522%257D%252C%2522clqzq7roc00033b6h868p1gcm%2522%253A%257B%2522id%2522%253A%2522clqzq7roc00033b6h868p1gcm%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqzq8nbm004g3b6huk53l74w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clqzq8nd5000mefjif7ta9ycz%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522id%2522%253A%2522clqzq9za0007h3b6hiuk5fxte%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqzq9za0007h3b6hiuk5fxte%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15.595238095238088%257D)

Steps:
1. At the sandbox link above, click on the row actions ellipsis.
2. Click the "Demo" list item button.
3. Notice the dialog briefly displays then disappears

### Current behavior

The issue arises when integrating dialogs, drawers, and other overlays/portals with the `DataGrid` component. However, within the `DataGrid`, these portals are not functioning as expected. The event propagation from portals seems to clash with the `DataGrid`'s event handling mechanisms, leading to erratic and unintended behaviors. I've seen a few issues reported that have been seemingly resolved or closed, but I'm still experiencing issues with the latest versions.

A workaround is instead of each `RowAction` being a self-contained component with its associated dialogs and drawers, I am forced to consider a less modular, more intermingled approach. This not only deviates from the planned architecture but also results in a less maintainable code structure.

### Expected behavior

The ideal scenario would involve portals working harmoniously within the `DataGrid` environment, allowing events to propagate correctly without disrupting the grid's functionality. This would enable the implementation of dialogs and drawers as part of individual RowAction components, maintaining the modular and organized code structure.

The `DataGrid` should support and encourage the development of modular, encapsulated components, in line with React and Material UI best practices. This would facilitate a clean, maintainable, and scalable codebase.

### Context

I am currently encountering an issue in the implementation of modular components within the `DataGrid`. The design philosophy I'm using hinges on the encapsulation of logic into self-contained, modular components. Each `RowAction`, such as in the sandbox example, is intended to be an autonomous component, encapsulating related drawers and dialogs. This approach aligns with modern React best practices, aiming for maintainable, organized, and reusable code.

However, I am facing a critical challenge due to the interaction between the portal implementation (for dialogs and drawers) and the event propagation behavior within the `DataGrid` component. This interaction is causing unexpected behaviors and inconsistencies, undermining the intended modularity of the components.

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  System:
    OS: macOS 14.2.1
  Binaries:
    Node: 18.16.0 - ~/.nvm/versions/node/v18.16.0/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn // not using Yarn, I'm using PNPM
    npm: 9.5.1 - ~/.nvm/versions/node/v18.16.0/bin/npm
  Browsers:
    Chrome: Not Found // using Brave
    Edge: Not Found
    Safari: 17.2.1
  npmPackages:
    @emotion/react: 11.11.1 => 11.11.1 
    @emotion/styled: 11.11.0 => 11.11.0 
    @mui/lab: 5.0.0-alpha.155 => 5.0.0-alpha.155 
    @mui/material: ^5.14.20 => 5.14.20 
    @mui/x-data-grid-generator: ^6.18.4 => 6.18.4 
    @mui/x-data-grid-premium: ^6.18.4 => 6.18.4 
    @mui/x-date-pickers: ^6.18.4 => 6.18.4 
    @mui/x-license-pro: ^6.10.2 => 6.10.2 
    @types/react: 18.2.43 => 18.2.43 
    react: 18.2.0 => 18.2.0 
    react-dom: 18.2.0 => 18.2.0 
    typescript: ~5.3.3 => 5.3.3 
```
</details>

**Search keywords**: datagrid gridcellactionitem portals events dialogs drawers
**Order ID**: 80458

**Comment by michelengelen** [comment]
Hey @irontitan76 ... Your assumption that it has something to do with events is unfortunately incorrect. The reason the dialog is disappearing is simply that it gets unmounted. In order for this to work you would need to extract the Dialog rendering to outside of the row.

Here you can see that the component gets unmounted: [example](https://codesandbox.io/p/devbox/mui-datagrid-portal-row-action-issue-forked-ghdj28?file=%2Fsrc%2Fdemo.tsx%3A1%2C1)

**Comment by irontitan76** [comment]
@michelengelen I investigated this a little bit before, and it seemed that events originating from the cell continue to fire even after the dialog gains focus. To address this, I intercepted and halted the propagation of these events. This approach successfully kept the dialog open, but it also led to unforeseen complications. While event bubbling and the component lifecycle are separate, they can intersect. Whether these issues stem from event bubbling—blurs/focuses,etc—that trigger component unmounting, or if they arise from an entirely different source, this is still an issue.

In the context of React, our goal is to develop self-contained, isolated components. This issue challenges that principle, as it necessitates relocating the dialog to sit adjacent to what would otherwise be its parent component. This workaround contradicts React's best practices of component modularity and hierarchy as well as component-driven development, indicating a deeper issue that needs to be addressed for a more sustainable and conforming solution.

TLDR; is the code in the sandbox valid code? If it is, why isn't it working as intended (as it should)?

**Comment by michelengelen** [comment]
Hi @irontitan76 I do understand this, but the way you have built it will (unintentionally) rerender the components. I did write up a quick example on how I would build this. It is arguably still modular :D.

[Example](https://codesandbox.io/p/devbox/mui-datagrid-portal-row-action-issue-forked-ghdj28?file=%2Fsrc%2Fdemo.tsx%3A15%2C20)

I understand that a component that would handle all of it in its own scope would be ideal, but given the nature of the actions its not working that way. This solution has some benefits as well and aims at using the dialog only once, but change its content based on the clicked action.

**Comment by irontitan76** [comment]
Thank you for supplying that example—it's similar to what I had to do when I first opened this issue. Regardless, I feel this bug should be resolved, especially for at-scale applications for the following reasons:

- In real world applications, a table or data grid often has several row actions.
- MUI X implemented the actions column definition to facilitate the former point, and the data grid API from MUI is one of the reasons many people use and pay for it.
- Within a table, each row action often has its own functionality, potentially with its own HTTP request.
- Extracting the row action and the ownership of the portal and the portal itself leads to more flexibility, modularity, and less messy and more readable code.
- Perhaps also, slotting within that sub-component  may be able to replace a trigger element—e.g., row action vs a normal button—for use elsewhere in the app, such as a details view, but the whole functionality is still retained. 
- It's less messy because I can define the request, manage the state, and the like within each isolated row action component.
 - If we had to do it the way suggested, worst-case, we likely have many state hooks and function calls defined: open for each portal, close for each portal, calls for content within portal, supplemental confirmation dialogs, and requests for numerous actions, and the like. Moreover, every row action is all within the same, now very messy, component.
- That said, some of—not all—functionality can reside in another component, but the open state and their handlers would still have to be defined within that messy component.
- The code I initially wrote is valid and is, perhaps arguably, follows best practices more closely, compared to the way we _had_ to write it.

So, I believe we should investigate specifically what is causing the re-render and fix it if it doesn’t have adverse effects to existing functionality. If it does, I feel we should at least provide an option to enable this more modular organization and development.

**Comment by michelengelen** [comment]
Ok, I get what you are saying and also what you are trying to do. The thing is: the menu is being rendered as a popper component and that will mount the content of the popover (menu) when opened and unmounted when closed. Therefore it's not possible with the provided features.

This is only true for the `GridActioncellItem`s that have `showInMenu` though, so any other are kept in the DOM and have more possibilities. This could possibly made more prominent in the docs (@joserodolfofreitas FYI) and I am sorry that your desired implementation is just not possible at the moment.

As for the reasoning I would defer to @romgrk or @cherniavskii for a more detailed explanation, but just imagine you have several millions of rows. Keeping a component with its own state (and possibly complex structure) for each of those rows can have a pretty massive hit on performance. There is virtualization in place as well, that's true, but depending on the size of the container and the amount of visible rows this can still hold true.

So I am not sure if we can support such a use-case, but as mentioned will defer this to the team as they have more in-depth knowledge about this.

**Comment by michelengelen** [comment]
@cherniavskii could you please check this? As mentioned above I think the main problem here is that the actions item is rendered in the menu and unmounted when the menu is hidden.

Thinking about this now: @irontitan76 would it help when you don't close the menu when clicking the actions item?

**Comment by cherniavskii** [comment]
I think adding the `hideMenuOnClick?: boolean` prop to `GridActionsCellItem` would make the most sense here.
@irontitan76 What do you think?
